handler/src/test/java/io/netty/handler/ssl/SslErrorTest.java

```diff
@@ -0,0 +1,248 @@
+                                @Override
+                                public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+                                    // Unwrap as its wrapped by a DecoderException
+                                    Throwable unwrappedCause = cause.getCause();
+                                    if (unwrappedCause instanceof SSLException) {
+                                        if (exception instanceof ValidatorException) {
+                                            CertPathValidatorException.Reason reason =
+                                                    ((CertPathValidatorException) exception.getCause()).getReason();
+                                            if (reason == CertPathValidatorException.BasicReason.EXPIRED) {
+                                                verifyException(unwrappedCause, "expired", promise);
+                                            } else if (reason == CertPathValidatorException.BasicReason.NOT_YET_VALID) {
+                                                verifyException(unwrappedCause, "bad", promise);
+                                            } else if (reason == CertPathValidatorException.BasicReason.REVOKED) {
+                                                verifyException(unwrappedCause, "revoked", promise);
+                                            }
+                                        } else if (exception instanceof CertificateExpiredException) {
+                                            verifyException(unwrappedCause, "expired", promise);
+                                        } else if (exception instanceof CertificateNotYetValidException) {
+                                            verifyException(unwrappedCause, "bad", promise);
+                                        } else if (exception instanceof CertificateRevokedException) {
+                                            verifyException(unwrappedCause, "revoked", promise);
+                                        }
+                                    }
+                                }
+                            });
+                        }
+                    }).connect(serverChannel.localAddress()).syncUninterruptibly().channel();
+            // Block until we received the correct exception
+            promise.syncUninterruptibly();
+        } finally {
+            if (clientChannel != null) {
+                clientChannel.close().syncUninterruptibly();
+            }
+            if (serverChannel != null) {
+                serverChannel.close().syncUninterruptibly();
+            }
+            group.shutdownGracefully();
+
+            ReferenceCountUtil.release(sslServerCtx);
+            ReferenceCountUtil.release(sslClientCtx);
+        }
+    }
+
+    // Its a bit hacky to verify against the message that is part of the exception but there is no other way
```

Comment:
agreed it is hacky ... maybe we can just validate the exception type for now and enhance later if JDK provides us with more granular exceptions, or a more reliable error code indicator?